# Notebook for developing preliminary models

In [1]:
# imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn

In [2]:
from sklearn.neighbors import KNeighborsRegressor
from sklearn.neural_network import MLPRegressor
from sklearn.svm import SVR
from sklearn.ensemble import RandomForestRegressor

In [3]:
# metric
from sklearn.metrics import mean_squared_log_error
def rmsle(y_true, y_pred, **kwargs):
    return np.sqrt(mean_squared_log_error(y_true, np.clip(y_pred, 0, None)))
rmsle_scorer = sklearn.metrics.make_scorer(rmsle, greater_is_better=False)

def rmsle_exp(y_true, y_pred, **kwargs):
    return np.sqrt(mean_squared_log_error(np.expm1(y_true), np.expm1(y_pred))) 
rmsle_scorer_exp = sklearn.metrics.make_scorer(rmsle_exp, greater_is_better=False)

In [4]:
from sklearn.model_selection import cross_val_score
def evaluate_CV(model, X, y, metric=rmsle_scorer, n_folds=5, random_state=None):
    return -cross_val_score(model, X, y, cv=n_folds, scoring=metric).mean()

## Data preparation for model training

### Importing data

In [5]:
# read in data Train data
train = pd.read_csv("train.csv", delimiter=",")
X_additional = pd.read_csv("additionalAttributes.csv", delimiter=",").drop("Unnamed: 0", axis=1)
X = train.drop(["formation_energy_ev_natom", "bandgap_energy_ev", "id"], axis=1)
y_fe = train.formation_energy_ev_natom
y_be = train.bandgap_energy_ev
X_full = pd.concat([X, X_additional], axis=1)

#Read in Test data
test = pd.read_csv("test.csv", delimiter=",")
X_test = test.drop(["id"], axis=1)
X_test_additional = pd.read_csv("additionalAttributesTest.csv", delimiter=",").drop("Unnamed: 0", axis=1)
X_full_Test = pd.concat([X_test, X_test_additional], axis=1)


In [6]:
X_full

spacegroup  number_of_total_atoms  percent_atom_al  percent_atom_ga  \
0             33                   80.0           0.6250           0.3750   
1            194                   80.0           0.6250           0.3750   
2            227                   40.0           0.8125           0.1875   
3            167                   30.0           0.7500           0.0000   
4            194                   80.0           0.0000           0.6250   
...          ...                    ...              ...              ...   
2395          33                   40.0           0.7500           0.2500   
2396         167                   30.0           0.4167           0.5833   
2397         206                   80.0           0.4375           0.5625   
2398          33                   80.0           0.3125           0.1875   
2399         206                   80.0           0.8750           0.0938   

      percent_atom_in  lattice_vector_1_ang  lattice_vector_2_ang  \
0              0.0000                9.9523                8.5513   
1              0.0000                6.1840                6.1838   
2              0.0000                9.7510                5.6595   
3              0.2500                5.0036                5.0034   
4              0.3750                6.6614                6.6612   
...               ...                   ...                   ...   
2395           0.0000                4.9469                8.5014   
2396           0.0000                4.9566                4.9562   
2397           0.0000                9.2204                9.2200   
2398           0.5000               10.6529                9.0954   
2399           0.0312                9.0648                9.0641   

      lattice_vector_3_ang  lattice_angle_alpha_degree  \
0                   9.1775                     90.0026   
1                  23.6287                     90.0186   
2                  13.9630                     90.9688   
3                  13.5318                     89.9888   
4                  24.5813                     89.9960   
...                    ...                         ...   
2395                9.1298                     90.0038   
2396               13.4178                     89.9938   
2397                9.2199                     90.0047   
2398                9.7210                     90.0015   
2399                9.0643                     90.0027   

      lattice_angle_beta_degree  ...  distGaO  distInAl  distInGa  distInO  \
0                       90.0023  ...    2.025       NaN       NaN      NaN   
1                       89.9980  ...    1.775     0.025     0.025    1.775   
2                       91.1228  ...    1.825     0.025     0.025    1.825   
3                       90.0119  ...    0.025     2.925     0.025    0.025   
4                       90.0006  ...    2.025     0.025     3.625    2.025   
...                         ...  ...      ...       ...       ...      ...   
2395                    90.0023  ...    1.775     0.025     0.025    1.775   
2396                    90.0075  ...    2.025     0.025     0.025    2.025   
2397                    90.0046  ...    1.975     0.025     0.025    1.975   
2398                    89.9996  ...    2.125     3.375     3.525    2.125   
2399                    90.0037  ...    1.975     3.425     4.575    1.975   

          elInt       qAl       qGa       qIn        qO       xEq  
0    -41.086176  0.349004  0.345264       NaN -0.231734  4.896453  
1    -38.554098  0.344690  0.347084       NaN -0.230392  4.905702  
2    -33.957596  0.343019  0.339820       NaN -0.228280  4.925995  
3    -35.763321  0.337616       NaN  0.375759 -0.231435  4.898196  
4    -36.645650       NaN  0.342521  0.359175 -0.232511  4.869110  
...         ...       ...       ...       ...       ...       ...  
2395 -39.262887  0.346869  0.345489       NaN -0.231016  4.905482  
2396 -35.323868  0.345140  0.342118       NaN -0.228918  4.928134  
2397 -

### Remove nan values From data

In [9]:
X_full.columns

Index(['spacegroup', 'number_of_total_atoms', 'percent_atom_al',
       'percent_atom_ga', 'percent_atom_in', 'lattice_vector_1_ang',
       'lattice_vector_2_ang', 'lattice_vector_3_ang',
       'lattice_angle_alpha_degree', 'lattice_angle_beta_degree',
       'lattice_angle_gamma_degree', 'Vatom', 'cAlGa', 'cAlIn', 'cAlO',
       'cGaAl', 'cGaIn', 'cGaO', 'cInAl', 'cInGa', 'cInO', 'distAlGa',
       'distAlIn', 'distAlO', 'distGaAl', 'distGaIn', 'distGaO', 'distInAl',
       'distInGa', 'distInO', 'elInt', 'qAl', 'qGa', 'qIn', 'qO', 'xEq'],
      dtype='object')

* If distance in Nan, it is reasonable to set them to very large value as the atoms are infinitely far from each other
* If Coordination number is Nan then it is reasonable to set it as 0, as no atoms are in vicinity of chosen atom
* If any of the charges have Nan values it is set to 0

In [10]:
fillValues = {'cAlGa':0, 
              'cAlIn':0, 
              'cAlO':0,
              'cGaAl':0, 
              'cGaIn':0, 
              'cGaO':0, 
              'cInAl':0, 
              'cInGa':0, 
              'cInO':0,
              'distAlGa':100,
              'distAlIn':100,
              'distAlO':100,
              'distGaAl':100,
              'distGaIn':100,
              'distGaO':100,
              'distInAl':100,
              'distInGa':100,
              'distInO':100,
              'qAl':0, 
              'qGa':0, 
              'qIn':0, 
              'qO':0}

In [11]:
X_full = X_full.fillna(value=fillValues)
X_full_Test = X_full_Test.fillna(value=fillValues)

### Categorize spacegroups

In [12]:
#Space groups possessing a point of inversion are termed centrosymmetric;
#ome space groups have no symmetry element that can change the handedness of an object; these are termed enantiomorphic space groups
def setSymmetry(X):
    centrosymmetric = []
    enantiomorphic = []
    groupEnantiomorph = [1]+list(np.arange(3,6,1))+list(np.arange(16,25,1))+list(np.arange(75,81,1))+list(np.arange(89,99,1))+list(np.arange(143,147,1))+list(np.arange(149,156,1))+list(np.arange(168,174,1))+list(np.arange(177,183,1))+list(np.arange(195,200,1))+list(np.arange(207,215,1))
    groupCentrosym = [2]+list(np.arange(10,16,1))+list(np.arange(47,75,1))+list(np.arange(83,89,1))+list(np.arange(123,143,1))+list(np.arange(147,149,1))+list(np.arange(162,168,1))+list(np.arange(175,177,1))+list(np.arange(191,195,1))+list(np.arange(200,207,1))+list(np.arange(221,231,1))
    for i in range(len(X)):
        if(X.spacegroup[i] in groupEnantiomorph):
            enantiomorphic+=[1]
            centrosymmetric+=[0]
        elif(X.spacegroup[i] in groupCentrosym):
            enantiomorphic+=[0]
            centrosymmetric+=[1]
        else:
            enantiomorphic+=[0]
            centrosymmetric+=[0]
    X["centroSym"] = centrosymmetric
    X["enantioMorph"] = enantiomorphic
    return X

In [13]:
# transform encoding
def encode_spacegroup(X):
    # 1-2 triclinic
    # 3-15 monoclinic
    # 16-74 orthorhombic
    # 75-142 tetragonal
    # 143-167 trigonal
    # 168-194 hexagonal
    # 195-230 cubic
    # [ 33 194 227 167 206  12] are the possible spacegroup values
    # onehot encode each separately
    return pd.get_dummies(X, columns=["spacegroup"])

In [14]:
X_full = setSymmetry(X_full)
X_full = encode_spacegroup(X_full)


X_full_Test = setSymmetry(X_full_Test)
X_full_Test = encode_spacegroup(X_full_Test)

In [15]:
X_full

number_of_total_atoms  percent_atom_al  percent_atom_ga  \
0                      80.0           0.6250           0.3750   
1                      80.0           0.6250           0.3750   
2                      40.0           0.8125           0.1875   
3                      30.0           0.7500           0.0000   
4                      80.0           0.0000           0.6250   
...                     ...              ...              ...   
2395                   40.0           0.7500           0.2500   
2396                   30.0           0.4167           0.5833   
2397                   80.0           0.4375           0.5625   
2398                   80.0           0.3125           0.1875   
2399                   80.0           0.8750           0.0938   

      percent_atom_in  lattice_vector_1_ang  lattice_vector_2_ang  \
0              0.0000                9.9523                8.5513   
1              0.0000                6.1840                6.1838   
2              0.0000                9.7510                5.6595   
3              0.2500                5.0036                5.0034   
4              0.3750                6.6614                6.6612   
...               ...                   ...                   ...   
2395           0.0000                4.9469                8.5014   
2396           0.0000                4.9566                4.9562   
2397           0.0000                9.2204                9.2200   
2398           0.5000               10.6529                9.0954   
2399           0.0312                9.0648                9.0641   

      lattice_vector_3_ang  lattice_angle_alpha_degree  \
0                   9.1775                     90.0026   
1                  23.6287                     90.0186   
2                  13.9630                     90.9688   
3                  13.5318                     89.9888   
4                  24.5813                     89.9960   
...                    ...                         ...   
2395                9.1298                     90.0038   
2396               13.4178                     89.9938   
2397                9.2199                     90.0047   
2398                9.7210                     90.0015   
2399                9.0643                     90.0027   

      lattice_angle_beta_degree  lattice_angle_gamma_degree  ...        qO  \
0                       90.0023                     90.0017  ... -0.231734   
1                       89.9980                    120.0025  ... -0.230392   
2                       91.1228                     30.5185  ... -0.228280   
3                       90.0119                    120.0017  ... -0.231435   
4                       90.0006                    119.9893  ... -0.232511   
...                         ...                         ...  ...       ...   
2395                    90.0023                     90.0015  ... -0.231016   
2396                    90.0075                    120.0007  ... -0.228918   
2397                    90.0046                     89.9954  ... -0.230109   
2398                    89.9996                     90.0004  ... -0.235314   
2399                    90.0037                     89.9987  ... -0.231068   

           xEq  centroSym  enantioMorph  spacegroup_12  spacegroup_33  \
0     4.896453          0             0              0              1   
1     4.905702          1             0              0              0   
2     4.925995          1             0              0              0   
3     4.898196          1             0              0              0   
4     4.869110          1             0              0              0   
...        ...        ...           ...            ...            ...   
2395  4.905482          0             0              0              1   
2396  4.928134          1             0              0              0   
2397  4.873503          1             0              0              0   
2398  4.842888          0       

## Scale attributes to from 0 to 1

In [17]:
len(X_full_Test.columns)

43

In [18]:
len(X_full.columns)

43

In [19]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaler.fit(X_full)
X_fullMinMax = scaler.transform(X_full)


scaler = MinMaxScaler()
scaler.fit(X_full_Test)
X_full_Test_MinMax = scaler.transform(X_full_Test)

### Scale using Standard scaler

In [20]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X_full)

X_fullStandard = scaler.transform(X_full)

### Separate into Training and validation

In [21]:
from sklearn.model_selection import train_test_split
from sklearn import metrics

trainSize = 0.85

############## Separate Train into Validation and Train for Formation energy ####################
#Without Scaling
#X_train_energ, X_val_energ, y_train_energ, y_val_energ = train_test_split(X_full, y_fe, train_size=trainSize, random_state=1)
#Min-Max Scaling
X_train_energ, X_val_energ, y_train_energ, y_val_energ = train_test_split(X_fullMinMax, y_fe, train_size=trainSize, random_state=1)
#Standard scaling
#X_train_energ, X_val_energ, y_train_energ, y_val_energ = train_test_split(X_fullStandard, y_fe, train_size=trainSize, random_state=1)


############## Separate Train into Validation and Train for Band Gap ####################
#Without Scaling
#X_train_gap, X_val_gap, y_train_gap, y_val_gap = train_test_split(X_full, y_be, train_size=trainSize, random_state=1)
#Min-Max Scaling
X_train_gap, X_val_gap, y_train_gap, y_val_gap = train_test_split(X_fullMinMax, y_be, train_size=trainSize, random_state=1)
#Standard scaling
#X_train_gap, X_val_gap, y_train_gap, y_val_gap = train_test_split(X_fullStandard, y_be, train_size=trainSize, random_state=1)


# Model training

## KNN

#### Formation energy

In [37]:
#################### KNN ##############################

results_df = pd.DataFrame(columns=['model',"Neigbors","Metric","weight",'trainError', 'valError','crossVal'])

for metr in ["manhattan", "minkowski"]:
    for i in range(3, 31, 2):
        for weight in ['uniform', 'distance']:
            knn = KNeighborsRegressor(n_neighbors=i, weights=weight, metric=metr).fit(X_train_energ, y_train_energ)

#            print("Neighbours " + str(i))
#            print("Metric " + metr)

            trainError = rmsle(y_train_energ, knn.predict(X_train_energ))
            valError = rmsle(y_val_energ, knn.predict(X_val_energ))
            cross_Val = evaluate_CV(knn, X_fullMinMax, y_fe)
            results_df = results_df.append({'model': 'KNN',"Neigbors":i,"Metric":metr,"weight":weight,"trainError":trainError, 'valError':valError, 'deltaErrors':abs(trainError-valError), 'crossVal':cross_Val}, ignore_index=True)


In [38]:
results_df.sort_values(by=["crossVal","valError", "deltaErrors"])

model Neigbors     Metric    weight  trainError  valError  crossVal  \
2    KNN        5  manhattan   uniform    0.026725  0.032796  0.033299   
4    KNN        7  manhattan   uniform    0.028354  0.031726  0.033543   
6    KNN        9  manhattan   uniform    0.029653  0.032113  0.033791   
9    KNN       11  manhattan  distance    0.006386  0.031603  0.033821   
7    KNN        9  manhattan  distance    0.006386  0.031832  0.033871   
8    KNN       11  manhattan   uniform    0.030733  0.032010  0.033873   
5    KNN        7  manhattan  distance    0.006386  0.031857  0.033888   
11   KNN       13  manhattan  distance    0.006386  0.031406  0.033929   
3    KNN        5  manhattan  distance    0.006386  0.033049  0.033945   
13   KNN       15  manhattan  distance    0.006386  0.031317  0.034028   
10   KNN       13  manhattan   uniform    0.031639  0.032023  0.034119   
15   KNN       17  manhattan  distance    0.006386  0.031421  0.034194   
17   KNN       19  manhattan  distance    0.006386  0.031726  0.034300   
12   KNN       15  manhattan   uniform    0.032155  0.032132  0.034321   
19   KNN       21  manhattan  distance    0.006386  0.031886  0.034421   
0    KNN        3  manhattan   uniform    0.024202  0.034312  0.034549   
21   KNN       23  manhattan  distance    0.006386  0.032094  0.034594   
14   KNN       17  manhattan   uniform    0.032611  0.032425  0.034621   
23   KNN       25  manhattan  distance    0.006386  0.032208  0.034786   
16   KNN       19  manhattan   uniform    0.032945  0.033045  0.034801   
34   KNN        9  minkowski   uniform    0.030833  0.033196  0.034807   
32   KNN        7  minkowski   uniform    0.029788  0.033164  0.034817   
25   KNN       27  manhattan  distance    0.006386  0.032285  0.034904   
18   KNN       21  manhattan   uniform    0.033424  0.033250  0.035034   
36   KNN       11  minkowski   uniform    0.031843  0.033511  0.035039   
35   KNN        9  minkowski  distance    0.006386  0.033019  0.035051   
30   KNN        5  minkowski   uniform    0.027999  0.034218  0.035079   
37   KNN       11  minkowski  distance    0.006386  0.033073  0.035120   
27   KNN       29  manhattan  distance    0.006386  0.032564  0.035136   
1    KNN        3  manhattan  distance    0.006386  0.034517  0.035181   
33   KNN        7  minkowski  distance    0.006386  0.033286  0.035193   
39   KNN       13  minkowski  distance    0.006386  0.033069  0.035284   
38   KNN       13  minkowski   uniform    0.032509  0.033661  0.035326   
20   KNN       23  manhattan   uniform    0.033725  0.033536  0.035418   
41   KNN       15  minkowski  distance    0.006386  0.032954  0.035446   
31   KNN        5  minkowski  distance    0.006386  0.034333  0.035604   
40   KNN       15  minkowski   uniform    0.033089  0.033722  0.035643   
43   KNN       17  minkowski  distance    0.006386  0.033068  0.035647   
22   KNN       25  manhattan   uniform    0.034118  0.033708  0.035735   
45   KNN       19  minkowski  distance    0.006386  0.033415  0.035763   
47   KNN       21  minkowski  distance    0.006386  0.033609  0.035909   
42   KNN       17  minkowski   uniform    0.033695  0.034061  0.035956   
24   KNN       27  manhattan   uniform    0.034507  0.033935  0.035980   
44   KNN       19  minkowski   uniform    0.034121  0.034586  0.036119   
49   KNN       23  minkowski  distance    0.006386  0.033761  0.036120   
51   KNN       25  minkowski  distance    0.006386  0.033895  0.036209   
28   KNN        3  minkowski   uniform    0.024773  0.035035  0.036264   
53   KNN       27  minkowski  distance    0.006386  0.034015  0.036336   
46   KNN       21  minkowski   uniform    0.034462  0.034878  0.036352   
26   KNN       29  manhattan   uniform    0.034944  0.034444  0.036429   
55   KNN       29  minkowski  distance    0.006386  0.034048  0.036480   
48   KNN       23  minkowski   uniform    0.034851  0.035102  0.036700   
29   KNN        3  minkowski  distance    0.006386  0.035101  0.036801   
50   KN

In [39]:
bestKnnFe = KNeighborsRegressor(n_neighbors=5, weights='uniform', metric='manhattan').fit(X_train_energ, y_train_energ)

KNN	5	manhattan	uniform	0.026725	0.032796	0.033299	0.006071

#### Band Gap

In [40]:
#################### KNN ##############################
results_df = pd.DataFrame(columns=['model',"Neigbors","Metric","weight",'trainError', 'valError'])

for metr in ["manhattan", "minkowski"]:
    for i in range(3, 31, 2):
        for weight in ['uniform', 'distance']:
            knn = KNeighborsRegressor(n_neighbors=i, weights=weight, metric=metr).fit(X_train_gap, y_train_gap)

#            print("Neighbours " + str(i))
#            print("Metric " + metr)

            trainError = rmsle(y_train_gap, knn.predict(X_train_gap))
            valError = rmsle(y_val_gap, knn.predict(X_val_gap))
            cross_Val = evaluate_CV(knn, X_fullMinMax, y_be)
            results_df = results_df.append({'model': 'KNN',"Neigbors":i,"Metric":metr,"weight":weight,"trainError":trainError, 'valError':valError, 'deltaErrors':abs(trainError-valError), 'crossVal':cross_Val}, ignore_index=True)


In [41]:
results_df.sort_values(by=["crossVal","valError", "deltaErrors"])

model Neigbors     Metric    weight  trainError  valError  crossVal  \
3    KNN        5  manhattan  distance    0.008884  0.086016  0.096120   
2    KNN        5  manhattan   uniform    0.076959  0.087256  0.097157   
7    KNN        9  manhattan  distance    0.008884  0.085056  0.097548   
5    KNN        7  manhattan  distance    0.008884  0.084873  0.097794   
9    KNN       11  manhattan  distance    0.008884  0.085306  0.098030   
11   KNN       13  manhattan  distance    0.008884  0.087785  0.099244   
1    KNN        3  manhattan  distance    0.008884  0.089373  0.099906   
0    KNN        3  manhattan   uniform    0.070379  0.091045  0.100146   
13   KNN       15  manhattan  distance    0.008884  0.089261  0.100634   
4    KNN        7  manhattan   uniform    0.083849  0.087318  0.100650   
6    KNN        9  manhattan   uniform    0.088817  0.088883  0.101815   
15   KNN       17  manhattan  distance    0.008884  0.091465  0.102217   
33   KNN        7  minkowski  distance    0.008884  0.090050  0.102260   
35   KNN        9  minkowski  distance    0.008884  0.092513  0.102972   
17   KNN       19  manhattan  distance    0.008884  0.093828  0.103320   
8    KNN       11  manhattan   uniform    0.092397  0.090753  0.103355   
37   KNN       11  minkowski  distance    0.008884  0.093680  0.103433   
31   KNN        5  minkowski  distance    0.008884  0.093879  0.103577   
30   KNN        5  minkowski   uniform    0.083385  0.094814  0.104730   
19   KNN       21  manhattan  distance    0.008884  0.095308  0.104738   
32   KNN        7  minkowski   uniform    0.087721  0.091650  0.104773   
39   KNN       13  minkowski  distance    0.008884  0.095317  0.104842   
10   KNN       13  manhattan   uniform    0.096406  0.095765  0.105653   
41   KNN       15  minkowski  distance    0.008884  0.097222  0.106037   
21   KNN       23  manhattan  distance    0.008884  0.097050  0.106171   
29   KNN        3  minkowski  distance    0.008884  0.099841  0.106375   
28   KNN        3  minkowski   uniform    0.073877  0.100882  0.106668   
34   KNN        9  minkowski   uniform    0.092914  0.096173  0.107115   
43   KNN       17  minkowski  distance    0.008884  0.098689  0.107230   
23   KNN       25  manhattan  distance    0.008884  0.098799  0.107275   
12   KNN       15  manhattan   uniform    0.099518  0.099197  0.108072   
36   KNN       11  minkowski   uniform    0.097202  0.099095  0.108332   
25   KNN       27  manhattan  distance    0.008884  0.100201  0.108516   
45   KNN       19  minkowski  distance    0.008884  0.100413  0.108729   
27   KNN       29  manhattan  distance    0.008884  0.102087  0.110110   
47   KNN       21  minkowski  distance    0.008884  0.101778  0.110131   
38   KNN       13  minkowski   uniform    0.100687  0.102288  0.110580   
14   KNN       17  manhattan   uniform    0.102620  0.102651  0.110726   
49   KNN       23  minkowski  distance    0.008884  0.103111  0.111479   
16   KNN       19  manhattan   uniform    0.104944  0.106313  0.112557   
51   KNN       25  minkowski  distance    0.008884  0.104761  0.112828   
40   KNN       15  minkowski   uniform    0.103885  0.106210  0.112914   
53   KNN       27  minkowski  distance    0.008884  0.106575  0.114443   
42   KNN       17  minkowski   uniform    0.106576  0.108750  0.114582   
18   KNN       21  manhattan   uniform    0.107620  0.108146  0.114756   
55   KNN       29  minkowski  distance    0.008884  0.109227  0.116001   
44   KNN       19  minkowski   uniform    0.109353  0.110250  0.116646   
20   KNN       23  manhattan   uniform    0.109502  0.110397  0.117147   
46   KNN       21  minkowski   uniform    0.111736  0.111854  0.118497   
22   KNN       25  manhattan   uniform    0.111900  0.112570  0.118932   
48   KNN       23  minkowski   uniform    0.113387  0.113350  0.120458   
24   KNN       27  manhattan   uniform    0.114018  0.114749  0.121119   
50   KNN       25  minkowski   uniform    0.115609  0.115256  0.122483   
26   KN

KNN	5	manhattan	distance	0.008884	0.086016	0.096120	0.077132

In [44]:
bestKnnGap = KNeighborsRegressor(n_neighbors=5, weights='distance', metric='manhattan').fit(X_train_gap, y_train_gap)

## SVM

#### Formation energy

In [46]:

################ SVR ##################################

kernels = ['poly', 'rbf']
gammas=['scale', 'auto']
for ker in kernels:
    for gam in gammas:
        svm = SVR(kernel=ker, gamma=gam).fit(X_train_energ, y_train_energ)
        trainError = rmsle(y_train_energ, abs(svm.predict(X_train_energ)))
        valError = rmsle(y_val_energ, abs(svm.predict(X_val_energ)))
        cross_Val = evaluate_CV(svm, X_fullMinMax, y_fe)
        print("Kernel: " + ker)
        print("Gamma: "+ gam)
        print("Training error: "+str(trainError))
        print("Validation error: "+ str(valError))
        print("Cross Val: "+ str(cross_Val))

Kernel: poly
Gamma: scale
Training error: 0.04575511989007159
Validation error: 0.046548858967812815
Cross Val: 0.0468249760291898
Kernel: poly
Gamma: auto
Training error: 0.05368859105287863
Validation error: 0.053315590896120375
Cross Val: 0.054080919423228756
Kernel: rbf
Gamma: scale
Training error: 0.04655858501114167
Validation error: 0.04670380970846281
Cross Val: 0.04760095925247003
Kernel: rbf
Gamma: auto
Training error: 0.04692385157175928
Validation error: 0.0462874440934817
Cross Val: 0.04790557016826317


Kernel: rbf
Gamma: scale
Training error: 0.04645671148891898
Validation error: 0.04642970080484417

#### Band gap

In [48]:

################ SVR ##################################
from sklearn.svm import SVR

kernels = ['poly', 'rbf']
gammas=['scale', 'auto']
for ker in kernels:
    for gam in gammas:
        svm = SVR(kernel=ker, gamma=gam).fit(X_train_gap, y_train_gap)
        trainError = rmsle(y_train_gap, abs(svm.predict(X_train_gap)))
        valError = rmsle(y_val_gap, abs(svm.predict(X_val_gap)))
        cross_Val = evaluate_CV(svm, X_fullMinMax, y_be)
        print("#########################")
        print("Kernel: " + ker)
        print("Gamma: "+ gam)
        print("Training error: "+str(trainError))
        print("Validation error: "+ str(valError))
        print("Cross Val: "+ str(cross_Val))


#########################
Kernel: poly
Gamma: scale
Training error: 0.07647495018153275
Validation error: 0.0714004739716549
Cross Val: 0.08448192991361378
#########################
Kernel: poly
Gamma: auto
Training error: 0.16001985199198773
Validation error: 0.16198828270740068
Cross Val: 0.16357661238014196
#########################
Kernel: rbf
Gamma: scale
Training error: 0.08178966104416957
Validation error: 0.07406221371391183
Cross Val: 0.08694838534130755
#########################
Kernel: rbf
Gamma: auto
Training error: 0.09706518792476887
Validation error: 0.08451282750292406
Cross Val: 0.09767289649953019


Kernel: poly
Gamma: scale
Training error: 0.07584564240632702
Validation error: 0.07156753875472792

### Random Forest

#### Formation energy

In [51]:
############### Random Forest ##############################

n_estimatorss = [650,750,850,1000]
max_depths=[30,35,40,45]
min_samples_splits=[4,6,8]
seeds = [1]
results_df = pd.DataFrame(columns=['model',"seed","n_estimators","max_depth","min_samples_split",'TrainError', 'ValError', 'deltaErrors'])

for seed in seeds:
    for n_estimator in n_estimatorss:
        for max_d in max_depths:
            for min_ss in min_samples_splits:
                rf = RandomForestRegressor(criterion='mse', n_estimators=n_estimator, max_depth=max_d, min_samples_split=min_ss, random_state=seed).fit(X_train_energ, y_train_energ)
                trainError = rmsle(y_train_energ, (rf.predict(X_train_energ)))
                valError = rmsle(y_val_energ, (rf.predict(X_val_energ)))

                cross_Val = evaluate_CV(rf, X_fullMinMax, y_fe)
                results_df = results_df.append({'model': 'RF',"seed":seed,"n_estimators":n_estimator,"max_depth":max_d,"min_samples_split":min_ss,"TrainError":trainError, 'ValError':valError, 'deltaErrors':abs(trainError-valError),'crossVal':cross_Val}, ignore_index=True)


KeyboardInterrupt: 

In [ ]:
results_df.sort_values(by=["crossVal","ValError", "deltaErrors"])

39	RF	1	1000	35	4	0.013313	0.030207	0.016894	0.031341

In [ ]:
bestRfFe = RandomForestRegressor(criterion='mse', n_estimators=1000, max_depth=35, min_samples_split=4, random_state=1).fit(X_train_energ, y_train_energ)

#### Band Gap

In [ ]:
############### Random Forest ##############################

n_estimatorss = [650,750,850,1000]
max_depths=[30,35,40,45,50]
min_samples_splits=[4,6,8]
seeds = [1]
results_df = pd.DataFrame(columns=['model',"seed","n_estimators","max_depth","min_samples_split",'TrainError', 'ValError', 'deltaErrors'])

for seed in seeds:
    for n_estimator in n_estimatorss:
        for max_d in max_depths:
            for min_ss in min_samples_splits:
                rf = RandomForestRegressor(criterion='mse', n_estimators=n_estimator, max_depth=max_d, min_samples_split=min_ss, random_state=seed).fit(X_train_gap, y_train_gap)
                trainError = rmsle(y_train_gap, (rf.predict(X_train_gap)))
                valError = rmsle(y_val_gap, (rf.predict(X_val_gap)))
                cross_Val = evaluate_CV(rf, X_fullMinMax, y_be)
                results_df = results_df.append({'model': 'RF',"seed":seed,"n_estimators":n_estimator,"max_depth":max_d,"min_samples_split":min_ss,"TrainError":trainError, 'ValError':valError, 'deltaErrors':abs(trainError-valError),'crossVal':cross_Val}, ignore_index=True)


In [33]:
results_df.sort_values(by=["crossVal","ValError", "deltaErrors"])

model seed n_estimators max_depth min_samples_split  TrainError  ValError  \
31    RF    1          850        30                 6    0.041993  0.080487   
34    RF    1          850        35                 6    0.041993  0.080487   
37    RF    1          850        40                 6    0.041993  0.080487   
40    RF    1          850        45                 6    0.041993  0.080487   
43    RF    1          850        50                 6    0.041993  0.080487   
33    RF    1          850        35                 4    0.038443  0.080184   
36    RF    1          850        40                 4    0.038443  0.080184   
39    RF    1          850        45                 4    0.038443  0.080184   
42    RF    1          850        50                 4    0.038443  0.080184   
30    RF    1          850        30                 4    0.038444  0.080191   
3     RF    1          650        35                 4    0.038497  0.080094   
6     RF    1          650        40                 4    0.038497  0.080094   
9     RF    1          650        45                 4    0.038497  0.080094   
12    RF    1          650        50                 4    0.038497  0.080094   
0     RF    1          650        30                 4    0.038498  0.080104   
16    RF    1          750        30                 6    0.041985  0.080497   
19    RF    1          750        35                 6    0.041985  0.080497   
22    RF    1          750        40                 6    0.041985  0.080497   
25    RF    1          750        45                 6    0.041985  0.080497   
28    RF    1          750        50                 6    0.041985  0.080497   
1     RF    1          650        30                 6    0.042023  0.080508   
4     RF    1          650        35                 6    0.042023  0.080508   
7     RF    1          650        40                 6    0.042023  0.080508   
10    RF    1          650        45                 6    0.042023  0.080508   
13    RF    1          650        50                 6    0.042023  0.080508   
18    RF    1          750        35                 4    0.038449  0.080156   
21    RF    1          750        40                 4    0.038449  0.080156   
24    RF    1          750        45                 4    0.038449  0.080156   
27    RF    1          750        50                 4    0.038449  0.080156   
15    RF    1          750        30                 4    0.038450  0.080164   
46    RF    1         1000        30                 6    0.042031  0.080457   
49    RF    1         1000        35                 6    0.042031  0.080459   
52    RF    1         1000        40                 6    0.042031  0.080459   
55    RF    1         1000        45                 6    0.042031  0.080459   
58    RF    1         1000        50                 6    0.042031  0.080459   
48    RF    1         1000        35                 4    0.038473  0.080171   
51    RF    1         1000        40                 4    0.038473  0.080171   
54    RF    1         1000        45                 4    0.038473  0.080171   
57    RF    1         1000        50                 4    0.038473  0.080171   
45    RF    1         1000        30                 4    0.038474  0.080178   
32    RF    1          850        30                 8    0.045452  0.080785   
35    RF    1          850        35                 8    0.045452  0.080785   
38    RF    1          850        40                 8    0.045452  0.080785   
41    RF    1          850        45                 8    0.045452  0.080785   
44    RF    1          850        50                 8    0.045452  0.080785   
17    RF    1          750        30                 8    0.045478  0.080811   
20    RF    1          750        35                 8    0.045478  0.080811   
23    RF    1          750        40                 8    0.045478  0.080811   
26    RF    1          750        45                 8    0.045478  0.080811   
29 

31	RF	1	850	30	6	0.041993	0.080487	0.038494	0.089325

In [ ]:
bestRfGap = RandomForestRegressor(criterion='mse', n_estimators=850, max_depth=30, min_samples_split=6, random_state=1).fit(X_train_gap, y_train_gap)

### Neutral Net (Multi-Layer Perceptron regressor)

#### Formation energy

In [35]:
############### ANN ##############################
hidden_layer_sizes = [(100,),(50,50), (50,100,50)]
activations = ["identity", "logistic", "tanh", "relu"]
solvers=['lbfgs','adam']
learning_rates = ["constant", "invscaling", "adaptive"]
seeds = [1]
results_df = pd.DataFrame(columns=['activation',"hidden_layer_size","solver","learning_rate",'TrainError', 'ValError', 'deltaErrors'])

for activation in activations:
    for solver in solvers:
        for learning_rate in learning_rates:
            for hidden_layer_size in hidden_layer_sizes:
                mlp = MLPRegressor(activation=activation, solver=solver, learning_rate=learning_rate, hidden_layer_sizes=hidden_layer_size).fit(X_train_energ, y_train_energ)
                trainError = rmsle(y_train_energ, abs(mlp.predict(X_train_energ)))
                valError = rmsle(y_val_energ, abs(mlp.predict(X_val_energ)))
                cross_Val = evaluate_CV(mlp, X_fullMinMax, y_fe)
                results_df = results_df.append({'activation':activation,"hidden_layer_size":hidden_layer_size,"solver":solver,"learning_rate":learning_rate,"TrainError":trainError, 'ValError':valError, 'deltaErrors':abs(trainError-valError),'crossVal':cross_Val}, ignore_index=True)


In [36]:
results_df.sort_values(by=["crossVal","ValError", "deltaErrors"])

activation hidden_layer_size solver learning_rate  TrainError  ValError  \
58       relu          (50, 50)  lbfgs    invscaling    0.030703  0.030816   
54       relu            (100,)  lbfgs      constant    0.026771  0.030554   
60       relu            (100,)  lbfgs      adaptive    0.027959  0.030064   
41       tanh     (50, 100, 50)  lbfgs    invscaling    0.030880  0.030808   
57       relu            (100,)  lbfgs    invscaling    0.027824  0.030575   
..        ...               ...    ...           ...         ...       ...   
35   logistic     (50, 100, 50)   adam      adaptive    0.052076  0.051262   
29   logistic     (50, 100, 50)   adam      constant    0.053991  0.052534   
23   logistic     (50, 100, 50)  lbfgs    invscaling    0.086962  0.085022   
26   logistic     (50, 100, 50)  lbfgs      adaptive    0.086831  0.084814   
20   logistic     (50, 100, 50)  lbfgs      constant    0.086695  0.084689   

    deltaErrors  crossVal  
58     0.000114  0.031325  
54     0.003783  0.031591  
60     0.002105  0.031689  
41     0.000072  0.031788  
57     0.002751  0.031805  
..          ...       ...  
35     0.000814  0.065256  
29     0.001457  0.066034  
23     0.001940  0.086544  
26     0.002017  0.086601  
20     0.002006  0.086615  

[72 rows x 8 columns]

58	relu	(50, 50)	lbfgs	invscaling	0.030703	0.030816	0.000114	0.031325

#### Band gap

In [37]:
############### ANN ##############################
hidden_layer_sizes = [(100,),(50,50), (50,100,50),(50,100,100,50)]
activations = ["identity", "logistic", "tanh", "relu"]
solvers=['lbfgs','adam']
learning_rates = ["constant", "invscaling", "adaptive"]
seeds = [1]
results_df = pd.DataFrame(columns=['activation',"hidden_layer_size","solver","learning_rate",'TrainError', 'ValError', 'deltaErrors'])

for activation in activations:
    for solver in solvers:
        for learning_rate in learning_rates:
            for hidden_layer_size in hidden_layer_sizes:
                mlp = MLPRegressor(activation=activation, solver=solver, learning_rate=learning_rate, hidden_layer_sizes=hidden_layer_size).fit(X_train_gap, y_train_gap)
                trainError = rmsle(y_train_gap, abs(mlp.predict(X_train_gap)))
                valError = rmsle(y_val_gap, abs(mlp.predict(X_val_gap)))
                cross_Val = evaluate_CV(mlp, X_fullMinMax, y_be)
                results_df = results_df.append({'activation':activation,"hidden_layer_size":hidden_layer_size,"solver":solver,"learning_rate":learning_rate,"TrainError":trainError, 'ValError':valError, 'deltaErrors':abs(trainError-valError),'crossVal':cross_Val}, ignore_index=True)


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, Converge

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:566: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, Converge

In [38]:
results_df.sort_values(by=["crossVal","ValError", "deltaErrors"])

activation hidden_layer_size solver learning_rate  TrainError  ValError  \
41       tanh     (50, 100, 50)  lbfgs    invscaling    0.076627  0.073540   
43       tanh          (50, 50)  lbfgs      adaptive    0.076890  0.075026   
38       tanh     (50, 100, 50)  lbfgs      constant    0.078058  0.071912   
44       tanh     (50, 100, 50)  lbfgs      adaptive    0.077632  0.072404   
40       tanh          (50, 50)  lbfgs    invscaling    0.077870  0.073002   
..        ...               ...    ...           ...         ...       ...   
9    identity            (100,)   adam      constant    0.113267  0.091627   
10   identity          (50, 50)   adam      constant    0.110986  0.090602   
13   identity          (50, 50)   adam    invscaling    0.114022  0.093414   
12   identity            (100,)   adam    invscaling    0.113628  0.092374   
14   identity     (50, 100, 50)   adam    invscaling    0.112729  0.089031   

    deltaErrors  crossVal  
41     0.003087  0.082592  
43     0.001864  0.082881  
38     0.006146  0.082995  
44     0.005228  0.083001  
40     0.004868  0.083208  
..          ...       ...  
9      0.021640  0.113185  
10     0.020384  0.113203  
13     0.020608  0.113265  
12     0.021254  0.114066  
14     0.023697  0.114144  

[72 rows x 8 columns]

41	tanh	(50, 100, 50)	lbfgs	invscaling	0.076627	0.073540	0.003087	0.082592

## Best models

Using the best trained models we form a weighted average ensemble

### Formation Energy

In [22]:
resultsFe = pd.DataFrame()

In [23]:
bestKnnFe = KNeighborsRegressor(n_neighbors=5, weights='uniform', metric='manhattan').fit(X_train_energ, y_train_energ)
cross_KNN_Fe = evaluate_CV(bestKnnFe, X_fullMinMax, y_fe)
valErrorKNN = rmsle(y_val_energ, (bestKnnFe.predict(X_val_energ)))
resultKNNFe = bestKnnFe.predict(X_full_Test_MinMax)
resultsFe["KNN"] = resultKNNFe
print(valErrorKNN)

0.03279555483040591


In [24]:
#Leaving this model out as this is noticeably worse model
#bestSvmFe = SVR(kernel="rbf", gamma="scale").fit(X_train_energ, y_train_energ)
#cross_SVM_Fe = evaluate_CV(bestSvmFe, X_fullMinMax, y_fe)
#valErrorSVM = rmsle(y_val_energ, (bestSvmFe.predict(X_val_energ)))
#resultSVMFe = bestSvmFe.predict(X_val_energ)
#resultsFe["SVM"] = resultSVMFe
#print(valErrorSVM)

In [25]:
bestRfFe = RandomForestRegressor(criterion='mse', n_estimators=1000, max_depth=35, min_samples_split=4, random_state=1).fit(X_train_energ, y_train_energ)
cross_RF_Fe = evaluate_CV(bestRfFe, X_fullMinMax, y_fe)
valErrorRF = rmsle(y_val_energ, (bestRfFe.predict(X_val_energ)))
resultRFFe = bestRfFe.predict(X_full_Test_MinMax)
resultsFe["RF"] = resultRFFe
print(valErrorRF)

0.030167500161596885


In [26]:
bestMlpFe = MLPRegressor(activation="relu", solver="lbfgs", learning_rate="invscaling", hidden_layer_sizes=(50, 50)).fit(X_train_energ, y_train_energ)
cross_Mlp_Fe = evaluate_CV(bestMlpFe, X_fullMinMax, y_fe)
valErrorMLP = rmsle(y_val_energ, (bestMlpFe.predict(X_val_energ)))
resultMlpFe = bestMlpFe.predict(X_full_Test_MinMax)
resultsFe["MLP"] = resultMlpFe
print(valErrorMLP)

0.03154900116513413


#### Use Cross-Ref error as weighting factor and evaluate Validation

In [27]:
summaryError = cross_KNN_Fe+cross_RF_Fe+cross_Mlp_Fe
resultsFe["Weighted"] = cross_KNN_Fe/summaryError*resultsFe["KNN"] + cross_RF_Fe/summaryError*resultsFe["RF"] + cross_Mlp_Fe/summaryError*resultsFe["MLP"]

#errorWeighted = rmsle(y_val_energ, resultsFe["Weighted"])

In [28]:
#print(errorWeighted)

### Band Gap

In [29]:
resultsGap = pd.DataFrame()

In [30]:
"""
bestKnnGap = KNeighborsRegressor(n_neighbors=5, weights='distance', metric='manhattan').fit(X_train_gap, y_train_gap)
cross_KNN_Gap = evaluate_CV(bestKnnGap, X_fullMinMax, y_be)

valErrorKNN = rmsle(y_val_gap, (bestKnnGap.predict(X_val_gap)))
resultKNNGap = bestKnnGap.predict(X_full_Test_MinMax)
resultsGap["KNN"] = resultKNNGap
print(valErrorKNN)
"""

'\nbestKnnGap = KNeighborsRegressor(n_neighbors=5, weights=\'distance\', metric=\'manhattan\').fit(X_train_gap, y_train_gap)\ncross_KNN_Gap = evaluate_CV(bestKnnGap, X_fullMinMax, y_be)\n\nvalErrorKNN = rmsle(y_val_gap, (bestKnnGap.predict(X_val_gap)))\nresultKNNGap = bestKnnGap.predict(X_full_Test_MinMax)\nresultsGap["KNN"] = resultKNNGap\nprint(valErrorKNN)\n'

In [31]:
bestSvmGap = SVR(kernel="poly", gamma="scale").fit(X_train_gap, y_train_gap)
cross_SVM_Gap = evaluate_CV(bestSvmGap, X_fullMinMax, y_be)

valErrorSVM = rmsle(y_val_gap, (bestSvmGap.predict(X_val_gap)))
resultSVMGap = bestSvmGap.predict(X_full_Test_MinMax)
resultsGap["SVM"] = resultSVMGap
print(valErrorSVM)

0.0714004739716549


In [32]:
"""
bestRfGap = RandomForestRegressor(criterion='mse', n_estimators=850, max_depth=30, min_samples_split=6, random_state=1).fit(X_train_gap, y_train_gap)
cross_RF_Gap = evaluate_CV(bestRfGap, X_fullMinMax, y_be)

valErrorRF = rmsle(y_val_gap, (bestRfGap.predict(X_val_gap)))
resultRFGap = bestRfGap.predict(X_full_Test_MinMax)
resultsGap["RF"] = resultRFGap
print(valErrorRF)
"""

'\nbestRfGap = RandomForestRegressor(criterion=\'mse\', n_estimators=850, max_depth=30, min_samples_split=6, random_state=1).fit(X_train_gap, y_train_gap)\ncross_RF_Gap = evaluate_CV(bestRfGap, X_fullMinMax, y_be)\n\nvalErrorRF = rmsle(y_val_gap, (bestRfGap.predict(X_val_gap)))\nresultRFGap = bestRfGap.predict(X_full_Test_MinMax)\nresultsGap["RF"] = resultRFGap\nprint(valErrorRF)\n'

In [33]:
len(X_full_Test_MinMax)

600

In [34]:
bestMlpGap = MLPRegressor(activation="tanh", solver="lbfgs", learning_rate="invscaling", hidden_layer_sizes=(50, 100, 50)).fit(X_train_gap, y_train_gap)
cross_Mlp_Gap = evaluate_CV(bestMlpFe, X_fullMinMax, y_be)

valErrorMlp = rmsle(y_val_gap, (bestMlpGap.predict(X_val_gap)))
resultMlpGap = bestMlpGap.predict(X_full_Test_MinMax)
resultsGap["MLP"] = resultMlpGap
print(valErrorMlp)

0.07282709000517473


#### Use Cross-Ref error as weighting factor and evaluate Validation

In [35]:
#summaryError = cross_SVM_Gap+cross_RF_Gap+cross_Mlp_Gap
#resultsGap["Weighted"] = cross_SVM_Gap/summaryError*resultsGap["SVM"] + cross_RF_Gap/summaryError*resultsGap["RF"] + cross_Mlp_Gap/summaryError*resultsGap["MLP"]

summaryError = cross_SVM_Gap+cross_Mlp_Gap
resultsGap["Weighted"] = cross_SVM_Gap/summaryError*resultsGap["SVM"] + cross_Mlp_Gap/summaryError*resultsGap["MLP"]


#errorWeighted = rmsle(y_val_gap, resultsGap["Weighted"])

In [36]:
#errorWeighted

In [37]:
# submission file
def save_results(y_fe_pred, y_be_pred, name):
    results = pd.DataFrame({"id": test.id, "formation_energy_ev_natom": y_fe_pred, "bandgap_energy_ev": y_be_pred})
    results.to_csv(name + ".csv",index=False)

In [39]:
save_results(resultsFe["RF"] ,resultsGap["SVM"], "output2")